### PR TITLE
fix(iroh): do not establish connection if content already exists locally

### LINKED ISSUE #1947: `blob_download` unnecessarily tries to connect to a peer even if the requested blob is found locally.
I ran into this issue when testing the RPC interface, by taking a `BlobTicket`, deconstructing it, constructing a `BlobDownloadRequest` and calling  `BlobsClient::download`. It's trivial to work around, simply call `BlobsClient::read` or `BlobsClient::read_to_bytes` first and use that result if you don't get an error. Perhaps this behaviour is intended, but it's a little unintuitive.

### PR DESCRIPTION
**Pull request by dignifiedquire**
Avoid establishing a connection in `blobs.download` if content is already locally available.

Closes #1947

### CODE REVIEW

**Review by flub — APPROVED** [review]

**Comment by dignifiedquire** [comment]
@Arqu @flub any ideas why this breaks netsim?

**Comment by Arqu** [comment]
Netsim output says it timed out & the logs don't really yell about anything. I guess it hangs for some reason?

TLDR; just tested this out manually. Transfer hangs. 
```
tty1
export IROH_DATA_DIR=/foo/blah-1
./target/release/iroh start --add ./target/release/iroh
# copy ticket
tty2
export IROH_DATA_DIR=/foo/blah-2
./target/release/iroh blob get --start blobticket --out STDOUT > /dev/null
```
This just hangs infinitely 
Screenshot 2024-01-23 at 12 43 25

**Review by rklaehn — APPROVED** [review]

File: iroh-bytes/src/get/db.rs [hunk 1]
```diff
@@ -33,26 +34,34 @@ use tracing::trace;
 ///
 /// This considers data that is already in the store, and will only request
 /// the remaining data.
-pub async fn get_to_db<D: BaoStore>(
+pub async fn get_to_db<
+    D: BaoStore,
+    C: FnOnce() -> F,
```
Comment (rklaehn) [inline iroh-bytes/src/get/db.rs:39]: :+1: for FnOnce here - you wouldn't want to spam tons of connections...
